In [1]:
%load_ext autoreload
%autoreload 2

In [58]:
import os
import pickle

import numpy as np
import pandas as pd

import settings as conf
import metadata

# Load S-MultiXcan results

In [3]:
smultixcan_pvalues_file = os.path.join(conf.GENE_ASSOC_DIR, f'smultixcan-mashr-pvalues.pkl.xz')
display(smultixcan_pvalues_file)

'/mnt/phenomexcan_base/gene_assoc/smultixcan-mashr-pvalues.pkl.xz'

In [4]:
smultixcan_gene_associations = pd.read_pickle(smultixcan_pvalues_file)

In [5]:
smultixcan_gene_associations.shape

(22515, 4091)

In [6]:
smultixcan_gene_associations.head(5)

4270-Volume_level_set_by_participant_left  \
gene_name                                                    
ENSG00000000419                                   0.952828   
ENSG00000000457                                   0.646365   
ENSG00000000460                                   0.896074   
ENSG00000000938                                   0.576924   
ENSG00000000971                                   0.956390   

                 S05-Diagnoses_main_ICD10_S05_Injury_of_eye_and_orbit  \
gene_name                                                               
ENSG00000000419                                           0.827955      
ENSG00000000457                                           0.299606      
ENSG00000000460                                           0.094218      
ENSG00000000938                                           0.107121      
ENSG00000000971                                           0.492012      

                 20003_1141157402-Treatmentmedication_code_prednisolone_product  \
gene_name                                                                         
ENSG00000000419                                           0.976745                
ENSG00000000457                                           0.477472                
ENSG00000000460                                           0.507646                
ENSG00000000938                                           0.439276                
ENSG00000000971                                           0.510924                

                 20002_1427-Noncancer_illness_code_selfreported_polycystic_kidney  \
gene_name                                                                           
ENSG00000000419                                           0.916643                  
ENSG00000000457                                           0.003570                  
ENSG00000000460                                           0.166442                  
ENSG00000000938                                           0.567582                  
ENSG00000000971                                           0.532389                  

                 110001-Invitation_to_complete_online_24hour_recall_dietary_questionnaire_acceptance  \
gene_name                                                                                              
ENSG00000000419                                           0.580172                                     
ENSG00000000457                                           0.905365                                     
ENSG00000000460                                           0.974545                                     
ENSG00000000938                                           0.768334                                     
ENSG00000000971                                           0.555313                                     

                 22617_2442-Job_SOC_coding_Social_workers  \
gene_name                                                   
ENSG00000000419                                  0.270858   
ENSG00000000457                                  0.845711   
ENSG00000000460                                  0.628910   
ENSG00000000938                                  0.203873   
ENSG00000000971                                  0.993563   

                 J93-Diagnoses_main_ICD10_J93_Pneumothorax  \
gene_name                                                    
ENSG00000000419                                   0.779171   
ENSG00000000457                                   0.137430   
ENSG00000000460                                   0.425188   
ENSG00000000938                                   0.231449   
ENSG00000000971                                   0.807439   

                 22601_41133206-Job_coding_local_government_administrative_officer_or_assistant_or_clerk  \
gene_name                                                                                                  
ENSG00000000419                                           0.597882                  

# Load fastENLOC results

In [7]:
fastenloc_rcp_file = os.path.join(conf.GENE_ASSOC_DIR, f'fastenloc-torus-rcp.pkl.xz')
display(fastenloc_rcp_file)

'/mnt/phenomexcan_base/gene_assoc/fastenloc-torus-rcp.pkl.xz'

In [8]:
fastenloc_gene_associations = pd.read_pickle(fastenloc_rcp_file)

In [9]:
fastenloc_gene_associations.shape

(37967, 4091)

In [10]:
fastenloc_gene_associations.head(5)

20485-Ever_contemplated_selfharm  \
gene_name                                           
ENSG00000000419                          0.000008   
ENSG00000000457                          0.000131   
ENSG00000000460                          0.000014   
ENSG00000000938                          0.000019   
ENSG00000000971                          0.000006   

                 6160_2-Leisuresocial_activities_Pub_or_social_club  \
gene_name                                                             
ENSG00000000419                                           0.000505    
ENSG00000000457                                           0.002876    
ENSG00000000460                                           0.000975    
ENSG00000000938                                           0.000449    
ENSG00000000971                                           0.000674    

                 6139_3-Gas_or_solidfuel_cookingheating_An_open_solid_fuel_fire_that_you_use_regularly_in_winter_time  \
gene_name                                                                                                               
ENSG00000000419                                           0.000305                                                      
ENSG00000000457                                           0.003605                                                      
ENSG00000000460                                           0.005453                                                      
ENSG00000000938                                           0.014776                                                      
ENSG00000000971                                           0.001918                                                      

                 L12_FOLLICULARNAS-Other_and_unspecified_follicular_disorders  \
gene_name                                                                       
ENSG00000000419                                       3.300140e-11              
ENSG00000000457                                       8.840290e-11              
ENSG00000000460                                       9.497296e-11              
ENSG00000000938                                       8.855560e-11              
ENSG00000000971                                       4.448820e-10              

                 2654_8-Nonbutter_spread_type_details_Other_low_or_reduced_fat_spread  \
gene_name                                                                               
ENSG00000000419                                           0.000039                      
ENSG00000000457                                           0.000203                      
ENSG00000000460                                           0.000147                      
ENSG00000000938                                           0.000091                      
ENSG00000000971                                           0.000057                      

                 22617_8139-Job_SOC_coding_Assemblers_and_routine_operatives_nec  \
gene_name                                                                          
ENSG00000000419                                       3.736700e-07                 
ENSG00000000457                                       7.714700e-07                 
ENSG00000000460                                       1.070900e-06                 
ENSG00000000938                                       6.395000e-07                 
ENSG00000000971                                       1.796000e-07                 

                 I9_CARDMPRI-Cardiomyopathies_Primaryintrinsic  \
gene_name                                                        
ENSG00000000419                                       0.000010   
ENSG00000000457                                       0.000195   
ENSG00000000460                                       0.000029   
ENSG00000000938                                       0.000090   
ENSG00000000971                                       0.000044   

                 20003_1141188442-Treatmentmedication_c

In [11]:
assert fastenloc_gene_associations.min().min() >= 0
assert fastenloc_gene_associations.max().max() <= 3

# S-MultiXcan

## Counts of significant associations

In [12]:
smultixcan_gene_associations.shape

(22515, 4091)

In [13]:
all_pvalues = pd.Series(smultixcan_gene_associations.values.flatten()).dropna()
display(all_pvalues.shape)

(91055810,)

In [14]:
all_pvalues.describe()

count     9.105581e+07
mean      4.839226e-01
std       2.934763e-01
min      9.999889e-321
25%       2.266266e-01
50%       4.794625e-01
75%       7.378647e-01
max       1.000000e+00
dtype: float64

In [15]:
(-np.log10(all_pvalues)).describe()

count    9.105581e+07
mean     4.990325e-01
std      1.258476e+00
min     -0.000000e+00
25%      1.320233e-01
50%      3.192453e-01
75%      6.446892e-01
max      3.200000e+02
dtype: float64

In [16]:
PVALUE_THRESHOLD = (0.05 / all_pvalues.shape[0])
display(PVALUE_THRESHOLD)

5.491137797796759e-10

In [17]:
PVALUE_THRESHOLD = 5.49e-10

In [18]:
hits = (all_pvalues < PVALUE_THRESHOLD).sum()
display(hits)
display((hits / all_pvalues.shape[0]) * 100)

72994

0.08016402248247531

In [19]:
hits = (all_pvalues < 0.01).sum()
display(hits)
display((hits / all_pvalues.shape[0]) * 100)

1809587

1.9873383148203285

# fastENLOC

## Numbers of significant genes also found by fastENLOC

In [20]:
common_genes = fastenloc_gene_associations.index.intersection(smultixcan_gene_associations.index)
display(common_genes)

Index(['ENSG00000000419', 'ENSG00000000457', 'ENSG00000000460',
       'ENSG00000000938', 'ENSG00000000971', 'ENSG00000001036',
       'ENSG00000001084', 'ENSG00000001167', 'ENSG00000001460',
       'ENSG00000001461',
       ...
       'ENSG00000284240', 'ENSG00000284308', 'ENSG00000284395',
       'ENSG00000284413', 'ENSG00000284418', 'ENSG00000284430',
       'ENSG00000284452', 'ENSG00000284513', 'ENSG00000284526',
       'ENSG00000284552'],
      dtype='object', name='gene_name', length=22427)

In [21]:
fastenloc_matrix_sorted = fastenloc_gene_associations.loc[common_genes, smultixcan_gene_associations.columns]
display(fastenloc_matrix_sorted.shape)
display(fastenloc_matrix_sorted.head())

(22427, 4091)

4270-Volume_level_set_by_participant_left  \
gene_name                                                    
ENSG00000000419                                   0.000005   
ENSG00000000457                                   0.000345   
ENSG00000000460                                   0.000175   
ENSG00000000938                                   0.000003   
ENSG00000000971                                   0.000221   

                 S05-Diagnoses_main_ICD10_S05_Injury_of_eye_and_orbit  \
gene_name                                                               
ENSG00000000419                                       1.736894e-09      
ENSG00000000457                                       3.936200e-10      
ENSG00000000460                                       8.839000e-10      
ENSG00000000938                                       5.624000e-10      
ENSG00000000971                                       9.193074e-11      

                 20003_1141157402-Treatmentmedication_code_prednisolone_product  \
gene_name                                                                         
ENSG00000000419                                       6.786450e-11                
ENSG00000000457                                       3.342258e-09                
ENSG00000000460                                       3.874230e-09                
ENSG00000000938                                       4.038900e-10                
ENSG00000000971                                       1.621935e-10                

                 20002_1427-Noncancer_illness_code_selfreported_polycystic_kidney  \
gene_name                                                                           
ENSG00000000419                                       2.790300e-11                  
ENSG00000000457                                       1.555570e-09                  
ENSG00000000460                                       2.791602e-10                  
ENSG00000000938                                       4.402530e-11                  
ENSG00000000971                                       8.354587e-11                  

                 110001-Invitation_to_complete_online_24hour_recall_dietary_questionnaire_acceptance  \
gene_name                                                                                              
ENSG00000000419                                           0.000014                                     
ENSG00000000457                                           0.000474                                     
ENSG00000000460                                           0.000217                                     
ENSG00000000938                                           0.000137                                     
ENSG00000000971                                           0.000213                                     

                 22617_2442-Job_SOC_coding_Social_workers  \
gene_name                                                   
ENSG00000000419                                  0.001053   
ENSG00000000457                                  0.001378   
ENSG00000000460                                  0.001331   
ENSG00000000938                                  0.001900   
ENSG00000000971                                  0.001095   

                 J93-Diagnoses_main_ICD10_J93_Pneumothorax  \
gene_name                                                    
ENSG00000000419                               1.904000e-07   
ENSG00000000457                               3.213490e-06   
ENSG00000000460                               2.925180e-06   
ENSG00000000938                               1.751500e-07   
ENSG00000000971                               1.944580e-06   

                 22601_41133206-Job_coding_local_government_administrative_officer_or_assistant_or_clerk  \
gene_name                                                                                                  
ENSG00000000419                                       9.100802e-12                  

In [22]:
multixcan_matrix_sorted = smultixcan_gene_associations.loc[common_genes]
display(multixcan_matrix_sorted.shape)
display(multixcan_matrix_sorted.head())

(22427, 4091)

4270-Volume_level_set_by_participant_left  \
gene_name                                                    
ENSG00000000419                                   0.952828   
ENSG00000000457                                   0.646365   
ENSG00000000460                                   0.896074   
ENSG00000000938                                   0.576924   
ENSG00000000971                                   0.956390   

                 S05-Diagnoses_main_ICD10_S05_Injury_of_eye_and_orbit  \
gene_name                                                               
ENSG00000000419                                           0.827955      
ENSG00000000457                                           0.299606      
ENSG00000000460                                           0.094218      
ENSG00000000938                                           0.107121      
ENSG00000000971                                           0.492012      

                 20003_1141157402-Treatmentmedication_code_prednisolone_product  \
gene_name                                                                         
ENSG00000000419                                           0.976745                
ENSG00000000457                                           0.477472                
ENSG00000000460                                           0.507646                
ENSG00000000938                                           0.439276                
ENSG00000000971                                           0.510924                

                 20002_1427-Noncancer_illness_code_selfreported_polycystic_kidney  \
gene_name                                                                           
ENSG00000000419                                           0.916643                  
ENSG00000000457                                           0.003570                  
ENSG00000000460                                           0.166442                  
ENSG00000000938                                           0.567582                  
ENSG00000000971                                           0.532389                  

                 110001-Invitation_to_complete_online_24hour_recall_dietary_questionnaire_acceptance  \
gene_name                                                                                              
ENSG00000000419                                           0.580172                                     
ENSG00000000457                                           0.905365                                     
ENSG00000000460                                           0.974545                                     
ENSG00000000938                                           0.768334                                     
ENSG00000000971                                           0.555313                                     

                 22617_2442-Job_SOC_coding_Social_workers  \
gene_name                                                   
ENSG00000000419                                  0.270858   
ENSG00000000457                                  0.845711   
ENSG00000000460                                  0.628910   
ENSG00000000938                                  0.203873   
ENSG00000000971                                  0.993563   

                 J93-Diagnoses_main_ICD10_J93_Pneumothorax  \
gene_name                                                    
ENSG00000000419                                   0.779171   
ENSG00000000457                                   0.137430   
ENSG00000000460                                   0.425188   
ENSG00000000938                                   0.231449   
ENSG00000000971                                   0.807439   

                 22601_41133206-Job_coding_local_government_administrative_officer_or_assistant_or_clerk  \
gene_name                                                                                                  
ENSG00000000419                                           0.597882                  

In [23]:
assert fastenloc_matrix_sorted.shape == multixcan_matrix_sorted.shape

### Using stringent pvalue threshold

In [24]:
display(PVALUE_THRESHOLD)

5.49e-10

In [25]:
multixcan_signif = (multixcan_matrix_sorted < PVALUE_THRESHOLD)
display(multixcan_signif.shape)
display(multixcan_signif.head())

(22427, 4091)

4270-Volume_level_set_by_participant_left  \
gene_name                                                    
ENSG00000000419                                      False   
ENSG00000000457                                      False   
ENSG00000000460                                      False   
ENSG00000000938                                      False   
ENSG00000000971                                      False   

                 S05-Diagnoses_main_ICD10_S05_Injury_of_eye_and_orbit  \
gene_name                                                               
ENSG00000000419                                              False      
ENSG00000000457                                              False      
ENSG00000000460                                              False      
ENSG00000000938                                              False      
ENSG00000000971                                              False      

                 20003_1141157402-Treatmentmedication_code_prednisolone_product  \
gene_name                                                                         
ENSG00000000419                                              False                
ENSG00000000457                                              False                
ENSG00000000460                                              False                
ENSG00000000938                                              False                
ENSG00000000971                                              False                

                 20002_1427-Noncancer_illness_code_selfreported_polycystic_kidney  \
gene_name                                                                           
ENSG00000000419                                              False                  
ENSG00000000457                                              False                  
ENSG00000000460                                              False                  
ENSG00000000938                                              False                  
ENSG00000000971                                              False                  

                 110001-Invitation_to_complete_online_24hour_recall_dietary_questionnaire_acceptance  \
gene_name                                                                                              
ENSG00000000419                                              False                                     
ENSG00000000457                                              False                                     
ENSG00000000460                                              False                                     
ENSG00000000938                                              False                                     
ENSG00000000971                                              False                                     

                 22617_2442-Job_SOC_coding_Social_workers  \
gene_name                                                   
ENSG00000000419                                     False   
ENSG00000000457                                     False   
ENSG00000000460                                     False   
ENSG00000000938                                     False   
ENSG00000000971                                     False   

                 J93-Diagnoses_main_ICD10_J93_Pneumothorax  \
gene_name                                                    
ENSG00000000419                                      False   
ENSG00000000457                                      False   
ENSG00000000460                                      False   
ENSG00000000938                                      False   
ENSG00000000971                                      False   

                 22601_41133206-Job_coding_local_government_administrative_officer_or_assistant_or_clerk  \
gene_name                                                                                                  
ENSG00000000419                                              False                  

In [26]:
fastenloc_signif = (fastenloc_matrix_sorted > 0.1)
display(fastenloc_signif.shape)
display(fastenloc_signif.head())

(22427, 4091)

4270-Volume_level_set_by_participant_left  \
gene_name                                                    
ENSG00000000419                                      False   
ENSG00000000457                                      False   
ENSG00000000460                                      False   
ENSG00000000938                                      False   
ENSG00000000971                                      False   

                 S05-Diagnoses_main_ICD10_S05_Injury_of_eye_and_orbit  \
gene_name                                                               
ENSG00000000419                                              False      
ENSG00000000457                                              False      
ENSG00000000460                                              False      
ENSG00000000938                                              False      
ENSG00000000971                                              False      

                 20003_1141157402-Treatmentmedication_code_prednisolone_product  \
gene_name                                                                         
ENSG00000000419                                              False                
ENSG00000000457                                              False                
ENSG00000000460                                              False                
ENSG00000000938                                              False                
ENSG00000000971                                              False                

                 20002_1427-Noncancer_illness_code_selfreported_polycystic_kidney  \
gene_name                                                                           
ENSG00000000419                                              False                  
ENSG00000000457                                              False                  
ENSG00000000460                                              False                  
ENSG00000000938                                              False                  
ENSG00000000971                                              False                  

                 110001-Invitation_to_complete_online_24hour_recall_dietary_questionnaire_acceptance  \
gene_name                                                                                              
ENSG00000000419                                              False                                     
ENSG00000000457                                              False                                     
ENSG00000000460                                              False                                     
ENSG00000000938                                              False                                     
ENSG00000000971                                              False                                     

                 22617_2442-Job_SOC_coding_Social_workers  \
gene_name                                                   
ENSG00000000419                                     False   
ENSG00000000457                                     False   
ENSG00000000460                                     False   
ENSG00000000938                                     False   
ENSG00000000971                                     False   

                 J93-Diagnoses_main_ICD10_J93_Pneumothorax  \
gene_name                                                    
ENSG00000000419                                      False   
ENSG00000000457                                      False   
ENSG00000000460                                      False   
ENSG00000000938                                      False   
ENSG00000000971                                      False   

                 22601_41133206-Job_coding_local_government_administrative_officer_or_assistant_or_clerk  \
gene_name                                                                                                  
ENSG00000000419                                              False                  

In [27]:
multixcan_hits = multixcan_signif.sum().sum()
display(multixcan_hits)
display(multixcan_hits / multixcan_signif.size)

72990

0.000795541245816283

In [28]:
fastenloc_hits = fastenloc_signif.sum().sum()
display(fastenloc_hits)
display(fastenloc_hits / fastenloc_signif.size)

292506

0.0031881160110801164

In [29]:
fastenloc_and_multixcan_signif = (multixcan_signif & fastenloc_signif)
assert fastenloc_and_multixcan_signif.shape == multixcan_signif.shape == fastenloc_signif.shape

In [30]:
fe_mu_sum = fastenloc_and_multixcan_signif.sum().sum()
display(fe_mu_sum)
_perc = fe_mu_sum / multixcan_hits
display(_perc)
display(f'{(_perc * 100):.2f}%')

22473

0.30789149198520344

'30.79%'

### Conclusions

72,990 significant associations with S-MultiXcan (after taking the set of genes in common with fastENLOC, otherwise it's 72,994).

From these, 22,473 associations are also colocalized (30.79%)

# Export significant associations as a table

### S-MultiXcan

In [31]:
smultixcan_results = multixcan_matrix_sorted.unstack()

In [32]:
smultixcan_results.shape

(91748857,)

In [33]:
smultixcan_results.head()

                                           gene_name      
4270-Volume_level_set_by_participant_left  ENSG00000000419    0.952828
                                           ENSG00000000457    0.646365
                                           ENSG00000000460    0.896074
                                           ENSG00000000938    0.576924
                                           ENSG00000000971    0.956390
dtype: float64

In [34]:
smultixcan_results.isna().sum()

1049005

In [35]:
smultixcan_results = smultixcan_results.dropna()

In [36]:
smultixcan_results.shape

(90699852,)

In [37]:
assert smultixcan_results.isna().sum() == 0

In [38]:
(smultixcan_results < PVALUE_THRESHOLD).sum()

72990

### fastENLOC

In [39]:
fastenloc_results = fastenloc_matrix_sorted.unstack()

In [40]:
fastenloc_results.shape

(91748857,)

In [41]:
fastenloc_results.head()

                                           gene_name      
4270-Volume_level_set_by_participant_left  ENSG00000000419    0.000005
                                           ENSG00000000457    0.000345
                                           ENSG00000000460    0.000175
                                           ENSG00000000938    0.000003
                                           ENSG00000000971    0.000221
dtype: float64

In [42]:
fastenloc_results.isna().sum()

1554545

In [43]:
fastenloc_results = fastenloc_results.dropna()

In [44]:
fastenloc_results.shape

(90194312,)

In [45]:
final_results = pd.DataFrame(index=smultixcan_results.index.copy(), columns=['smultixcan_pval', 'fastenloc_rcp'])

In [46]:
final_results = final_results.assign(smultixcan_pval=smultixcan_results, fastenloc_rcp=fastenloc_results)

In [47]:
final_results.shape

(90699852, 2)

In [48]:
final_results.head()

smultixcan_pval  \
                                          gene_name                          
4270-Volume_level_set_by_participant_left ENSG00000000419         0.952828   
                                          ENSG00000000457         0.646365   
                                          ENSG00000000460         0.896074   
                                          ENSG00000000938         0.576924   
                                          ENSG00000000971         0.956390   

                                                           fastenloc_rcp  
                                          gene_name                       
4270-Volume_level_set_by_participant_left ENSG00000000419       0.000005  
                                          ENSG00000000457       0.000345  
                                          ENSG00000000460       0.000175  
                                          ENSG00000000938       0.000003  
                                          ENSG00000000971       0.000221

In [49]:
final_results.isna().sum()

smultixcan_pval          0
fastenloc_rcp      1529981
dtype: int64

In [50]:
final_results = final_results.dropna(how='any')

In [51]:
final_results.shape

(89169871, 2)

In [52]:
PVALUE_THRESHOLD

5.49e-10

In [53]:
publishable_final_results = final_results[(final_results['smultixcan_pval'] < PVALUE_THRESHOLD) & (final_results['fastenloc_rcp'] > 0.1)]

In [54]:
publishable_final_results.index.rename(('trait', 'gene_id'), inplace=True)

In [55]:
publishable_final_results.shape

(22473, 2)

In [56]:
publishable_final_results.head()

smultixcan_pval  \
trait                                              gene_id                            
C43-Diagnoses_main_ICD10_C43_Malignant_melanoma... ENSG00000101440     1.808977e-10   
                                                   ENSG00000204991     3.540612e-14   
M72-Diagnoses_main_ICD10_M72_Fibroblastic_disor... ENSG00000104408     7.674021e-24   
                                                   ENSG00000147655     2.652440e-26   
                                                   ENSG00000172590     2.479003e-11   

                                                                    fastenloc_rcp  
trait                                              gene_id                         
C43-Diagnoses_main_ICD10_C43_Malignant_melanoma... ENSG00000101440       0.970837  
                                                   ENSG00000204991       0.401100  
M72-Diagnoses_main_ICD10_M72_Fibroblastic_disor... ENSG00000104408       0.192500  
                                                   ENSG00000147655       0.103500  
                                                   ENSG00000172590       0.359586

In [60]:
metadata.GENES_MAPPINGS.head()

gene gene_name       gene_type          gene_id    band
0  ENSG00000000457.13     SCYL3  protein_coding  ENSG00000000457  1q24.2
1  ENSG00000000460.16  C1orf112  protein_coding  ENSG00000000460  1q24.2
2  ENSG00000000938.12       FGR  protein_coding  ENSG00000000938  1p35.3
3  ENSG00000000971.15       CFH  protein_coding  ENSG00000000971  1q31.3
4  ENSG00000001036.13     FUCA2  protein_coding  ENSG00000001036  6q24.2

In [61]:
_tmp = pd.merge(
    publishable_final_results.reset_index(),
    metadata.GENES_MAPPINGS[['gene_id', 'gene_name', 'band', 'gene_type']],
    on='gene_id',
    how='left'
).rename(columns={'band': 'gene_band'})

In [62]:
_tmp.head()

trait          gene_id  \
0  C43-Diagnoses_main_ICD10_C43_Malignant_melanom...  ENSG00000101440   
1  C43-Diagnoses_main_ICD10_C43_Malignant_melanom...  ENSG00000204991   
2  M72-Diagnoses_main_ICD10_M72_Fibroblastic_diso...  ENSG00000104408   
3  M72-Diagnoses_main_ICD10_M72_Fibroblastic_diso...  ENSG00000147655   
4  M72-Diagnoses_main_ICD10_M72_Fibroblastic_diso...  ENSG00000172590   

   smultixcan_pval  fastenloc_rcp gene_name gene_band       gene_type  
0     1.808977e-10       0.970837      ASIP  20q11.22  protein_coding  
1     3.540612e-14       0.401100    SPIRE2   16q24.3  protein_coding  
2     7.674021e-24       0.192500     EIF3E    8q23.1  protein_coding  
3     2.652440e-26       0.103500     RSPO2    8q23.1  protein_coding  
4     2.479003e-11       0.359586    MRPL52   14q11.2  protein_coding

In [63]:
publishable_final_results = _tmp.set_index(['trait', 'gene_id'])[['gene_name', 'gene_band', 'gene_type', 'smultixcan_pval', 'fastenloc_rcp']]

In [64]:
publishable_final_results = publishable_final_results.sort_index()

In [65]:
assert publishable_final_results.index.is_unique

In [66]:
publishable_final_results.isna().sum()

gene_name          11
gene_band          11
gene_type          11
smultixcan_pval     0
fastenloc_rcp       0
dtype: int64

In [67]:
publishable_final_results[publishable_final_results['gene_name'].isna()]

gene_name  \
trait                                              gene_id                     
23101_raw-Whole_body_fatfree_mass                  ENSG00000253972       NaN   
23102_raw-Whole_body_water_mass                    ENSG00000253972       NaN   
23105_raw-Basal_metabolic_rate                     ENSG00000253972       NaN   
23121_raw-Arm_fatfree_mass_right                   ENSG00000253972       NaN   
23122_raw-Arm_predicted_mass_right                 ENSG00000253972       NaN   
23125_raw-Arm_fatfree_mass_left                    ENSG00000253972       NaN   
23126_raw-Arm_predicted_mass_left                  ENSG00000253972       NaN   
23129_raw-Trunk_fatfree_mass                       ENSG00000253972       NaN   
23130_raw-Trunk_predicted_mass                     ENSG00000253972       NaN   
30090_raw-Platelet_crit                            ENSG00000229694       NaN   
4079_raw-Diastolic_blood_pressure_automated_rea... ENSG00000253972       NaN   

                                                                   gene_band  \
trait                                              gene_id                     
23101_raw-Whole_body_fatfree_mass                  ENSG00000253972       NaN   
23102_raw-Whole_body_water_mass                    ENSG00000253972       NaN   
23105_raw-Basal_metabolic_rate                     ENSG00000253972       NaN   
23121_raw-Arm_fatfree_mass_right                   ENSG00000253972       NaN   
23122_raw-Arm_predicted_mass_right                 ENSG00000253972       NaN   
23125_raw-Arm_fatfree_mass_left                    ENSG00000253972       NaN   
23126_raw-Arm_predicted_mass_left                  ENSG00000253972       NaN   
23129_raw-Trunk_fatfree_mass                       ENSG00000253972       NaN   
23130_raw-Trunk_predicted_mass                     ENSG00000253972       NaN   
30090_raw-Platelet_crit                            ENSG00000229694       NaN   
4079_raw-Diastolic_blood_pressure_automated_rea... ENSG00000253972       NaN   

                                                                   gene_type  \
trait                                              gene_id                     
23101_raw-Whole_body_fatfree_mass                  ENSG00000253972       NaN   
23102_raw-Whole_body_water_mass                    ENSG00000253972       NaN   
23105_raw-Basal_metabolic_rate                     ENSG00000253972       NaN   
23121_raw-Arm_fatfree_mass_right                   ENSG00000253972       NaN   
23122_raw-Arm_predicted_mass_right                 ENSG00000253972       NaN   
23125_raw-Arm_fatfree_mass_left                    ENSG00000253972       NaN   
23126_raw-Arm_predicted_mass_left                  ENSG00000253972       NaN   
23129_raw-Trunk_fatfree_mass                       ENSG00000253972       NaN   
23130_raw-Trunk_predicted_mass                     ENSG00000253972       NaN   
30090_raw-Platelet_crit                            ENSG00000229694       NaN   
4079_raw-Diastolic_blood_pressure_automated_rea... ENSG00000253972       NaN   

                                                                    smultixcan_pval  \
trait                                              gene_id                            
23101_raw-Whole_body_fatfree_mass                  ENSG00000253972     1.246210e-13   
23102_raw-Whole_body_water_mass                    ENSG00000253972     1.868528e-13   
23105_raw-Basal_metabolic_rate                     ENSG00000253972     6.701070e-11   
23121_raw-Arm_fatfree_mass_right                   ENSG00000253972     1.262021e-10   
23122_raw-Arm_predicted_mass_right                 ENSG00000253972     1.539772e-10   
23125_raw-Arm_fatfree_mass_left                    ENSG00000253972     1.553128e-11   
23126_raw-Arm_predicted_mass_left                  ENSG00000253972     2.576215e-11   
23129_raw-Trunk_fatfree_mass                       ENSG00000253972     3.512784e-15   
23130_raw-Trunk_predicted_mass                     ENSG00000253972     5.3

In [68]:
publishable_final_results.head()

gene_name gene_band  \
trait                                gene_id                               
102_raw-Pulse_rate_automated_reading ENSG00000031003    FAM13B    5q31.2   
                                     ENSG00000064989    CALCRL    2q32.1   
                                     ENSG00000072195      SPEG      2q35   
                                     ENSG00000075420    FNDC3B   3q26.31   
                                     ENSG00000075702     WDR62  19q13.12   

                                                           gene_type  \
trait                                gene_id                           
102_raw-Pulse_rate_automated_reading ENSG00000031003  protein_coding   
                                     ENSG00000064989  protein_coding   
                                     ENSG00000072195  protein_coding   
                                     ENSG00000075420  protein_coding   
                                     ENSG00000075702  protein_coding   

                                                      smultixcan_pval  \
trait                                gene_id                            
102_raw-Pulse_rate_automated_reading ENSG00000031003     9.316678e-29   
                                     ENSG00000064989     1.553027e-22   
                                     ENSG00000072195     1.707185e-38   
                                     ENSG00000075420     2.695478e-24   
                                     ENSG00000075702     8.858931e-11   

                                                      fastenloc_rcp  
trait                                gene_id                         
102_raw-Pulse_rate_automated_reading ENSG00000031003       0.868600  
                                     ENSG00000064989       0.863200  
                                     ENSG00000072195       0.877000  
                                     ENSG00000075420       0.841100  
                                     ENSG00000075702       1.257727

In [69]:
publishable_final_results.shape

(22473, 5)

In [71]:
# some testing
_result = publishable_final_results.loc[('1180-Morningevening_person_chronotype', metadata.GENE_NAME_TO_ID_MAP['VIP'])]
assert _result['smultixcan_pval'] == 1.8124965657173678e-17
assert _result['fastenloc_rcp'] == 0.30421

In [72]:
# some testing
_result = publishable_final_results.loc[('1180-Morningevening_person_chronotype', metadata.GENE_NAME_TO_ID_MAP['RP11-220I1.5'])]
assert _result['smultixcan_pval'] == 6.426619359444203e-11
assert _result['fastenloc_rcp'] == 0.2150000007971

In [75]:
# for publication (tsv.gz)
output_folder = os.path.join(conf.DELIVERABLES_DIR, 'supp_tables')
os.makedirs(output_folder, exist_ok=True)
output_file = os.path.join(output_folder, 'suppl_table_S1-significant_gene_trait_associations.tsv.gz')
display(output_file)

publishable_final_results.to_csv(output_file, sep='\t', float_format='%.4e')

'/mnt/phenomexcan_base/deliverables/supp_tables/suppl_table_S1-significant_gene_trait_associations.tsv.gz'

In [76]:
# some testing
_tmp = pd.read_csv(output_file, sep='\t')

In [77]:
assert publishable_final_results.shape == _tmp.set_index(['trait', 'gene_id']).shape

In [78]:
_tmp.shape

(22473, 7)

In [79]:
_tmp.head()

trait          gene_id gene_name gene_band  \
0  102_raw-Pulse_rate_automated_reading  ENSG00000031003    FAM13B    5q31.2   
1  102_raw-Pulse_rate_automated_reading  ENSG00000064989    CALCRL    2q32.1   
2  102_raw-Pulse_rate_automated_reading  ENSG00000072195      SPEG      2q35   
3  102_raw-Pulse_rate_automated_reading  ENSG00000075420    FNDC3B   3q26.31   
4  102_raw-Pulse_rate_automated_reading  ENSG00000075702     WDR62  19q13.12   

        gene_type  smultixcan_pval  fastenloc_rcp  
0  protein_coding     9.316700e-29         0.8686  
1  protein_coding     1.553000e-22         0.8632  
2  protein_coding     1.707200e-38         0.8770  
3  protein_coding     2.695500e-24         0.8411  
4  protein_coding     8.858900e-11         1.2577

In [80]:
_tmp = _tmp.set_index(['trait', 'gene_id'])

In [82]:
_result = _tmp.loc[('1180-Morningevening_person_chronotype', metadata.GENE_NAME_TO_ID_MAP['VIP'])]
assert _result['smultixcan_pval'] == 1.8125e-17
assert _result['fastenloc_rcp'] == 0.30421

_result = _tmp.loc[('1180-Morningevening_person_chronotype', metadata.GENE_NAME_TO_ID_MAP['RP11-220I1.5'])]
assert _result['smultixcan_pval'] == 6.4266e-11
assert _result['fastenloc_rcp'] == 0.2150

In [83]:
assert np.allclose(
    _tmp[['smultixcan_pval', 'fastenloc_rcp']].values,
    publishable_final_results[['smultixcan_pval', 'fastenloc_rcp']].values,
    atol=1e-320,
    rtol=1e-4
)

In [85]:
# for publication (xlsx)
output_folder = os.path.join(conf.DELIVERABLES_DIR, 'supp_tables')
os.makedirs(output_folder, exist_ok=True)
output_file = os.path.join(output_folder, 'suppl_table_S1-significant_gene_trait_associations.xlsx')
display(output_file)

publishable_final_results.reset_index().to_excel(output_file, index=False, float_format='%.4e')

'/mnt/phenomexcan_base/deliverables/supp_tables/suppl_table_S1-significant_gene_trait_associations.xlsx'

In [86]:
# some testing
_tmp = pd.read_excel(output_file)

In [87]:
assert publishable_final_results.shape == _tmp.set_index(['trait', 'gene_id']).shape

In [88]:
_tmp.shape

(22473, 7)

In [89]:
_tmp.head()

trait          gene_id gene_name gene_band  \
0  102_raw-Pulse_rate_automated_reading  ENSG00000031003    FAM13B    5q31.2   
1  102_raw-Pulse_rate_automated_reading  ENSG00000064989    CALCRL    2q32.1   
2  102_raw-Pulse_rate_automated_reading  ENSG00000072195      SPEG      2q35   
3  102_raw-Pulse_rate_automated_reading  ENSG00000075420    FNDC3B   3q26.31   
4  102_raw-Pulse_rate_automated_reading  ENSG00000075702     WDR62  19q13.12   

        gene_type  smultixcan_pval  fastenloc_rcp  
0  protein_coding     9.316700e-29         0.8686  
1  protein_coding     1.553000e-22         0.8632  
2  protein_coding     1.707200e-38         0.8770  
3  protein_coding     2.695500e-24         0.8411  
4  protein_coding     8.858900e-11         1.2577

In [90]:
_tmp = _tmp.set_index(['trait', 'gene_id'])

In [92]:
_result = _tmp.loc[('1180-Morningevening_person_chronotype', metadata.GENE_NAME_TO_ID_MAP['VIP'])]
assert _result['smultixcan_pval'] == 1.8125e-17
assert _result['fastenloc_rcp'] == 0.30421

_result = _tmp.loc[('1180-Morningevening_person_chronotype', metadata.GENE_NAME_TO_ID_MAP['RP11-220I1.5'])]
assert _result['smultixcan_pval'] == 6.4266e-11
assert _result['fastenloc_rcp'] == 0.2150

In [93]:
assert np.allclose(
    _tmp[['smultixcan_pval', 'fastenloc_rcp']].values,
    publishable_final_results[['smultixcan_pval', 'fastenloc_rcp']].values,
    atol=1e-320,
    rtol=1e-4
)

# QQ plots

In [ ]:
import plots
from clustering.biclustering.analysis import Trait

# Cholesterol

In [ ]:
spredixcan_genes_associations.columns[spredixcan_genes_associations.columns.str.lower().str.contains('cholesterol')]

In [ ]:
pheno_plain_name = '20002_1473-Noncancer_illness_code_selfreported_high_cholesterol'

In [ ]:
pheno_pvalues = pd.Series(spredixcan_genes_associations[pheno_plain_name].values.flatten())
display(pheno_pvalues.shape)

In [ ]:
threshold = 1e-30

unif_values = np.linspace(0.0001, 1, num=pheno_pvalues.shape[0])
unif_values[unif_values < threshold] = threshold

real_values = pheno_pvalues
real_values[real_values < threshold] = threshold

df = pd.DataFrame({'unif': unif_values, 'real': real_values})
display(df.shape)

In [ ]:
df_plot = df

ax = plots.qqplot(-np.log10(df_plot), 'unif', 'real', xlabel='Expected $-\log_{10}(p)$', ylabel='Observed $-\log_{10}(p)$', s=5, linewidth=0)
ax.set_xlim([0, 5])
ax.set_title(Trait(pheno_plain_name).get_pretty_name())

# Height

In [ ]:
spredixcan_genes_associations.columns[spredixcan_genes_associations.columns.str.lower().str.contains('height')]

In [ ]:
pheno_plain_name = '50_raw-Standing_height'

In [ ]:
pheno_pvalues = pd.Series(spredixcan_genes_associations[pheno_plain_name].values.flatten())
display(pheno_pvalues.shape)

In [ ]:
threshold = 1e-30

unif_values = np.linspace(0.0001, 1, num=pheno_pvalues.shape[0])
unif_values[unif_values < threshold] = threshold

real_values = pheno_pvalues
real_values[real_values < threshold] = threshold

df = pd.DataFrame({'unif': unif_values, 'real': real_values})
display(df.shape)

In [ ]:
df_plot = df

ax = plots.qqplot(-np.log10(df_plot), 'unif', 'real', xlabel='Expected $-\log_{10}(p)$', ylabel='Observed $-\log_{10}(p)$', s=5, linewidth=0)
ax.set_xlim([0, 5])
ax.set_title(Trait(pheno_plain_name).get_pretty_name())

# Schizophrenia

In [ ]:
spredixcan_genes_associations.columns[spredixcan_genes_associations.columns.str.lower().str.contains('schizophrenia')]

In [ ]:
pheno_plain_name = '20002_1289-Noncancer_illness_code_selfreported_schizophrenia'

In [ ]:
pheno_pvalues = pd.Series(spredixcan_genes_associations[pheno_plain_name].values.flatten())
display(pheno_pvalues.shape)

In [ ]:
threshold = 1e-30

unif_values = np.linspace(0.0001, 1, num=pheno_pvalues.shape[0])
unif_values[unif_values < threshold] = threshold

real_values = pheno_pvalues
real_values[real_values < threshold] = threshold

df = pd.DataFrame({'unif': unif_values, 'real': real_values})
display(df.shape)

In [ ]:
df_plot = df

ax = plots.qqplot(-np.log10(df_plot), 'unif', 'real', xlabel='Expected $-\log_{10}(p)$', ylabel='Observed $-\log_{10}(p)$', s=5, linewidth=0)
ax.set_xlim([0, 5])
ax.set_title(Trait(pheno_plain_name).get_pretty_name())

# All p-values

In [ ]:
threshold = 1e-30

unif_values = np.linspace(0.0001, 1, num=all_pvalues.shape[0])
unif_values[unif_values < threshold] = threshold

real_values = all_pvalues
real_values[real_values < threshold] = threshold

df = pd.DataFrame({'unif': unif_values, 'real': real_values})

In [ ]:
df_plot = df

ax = plots.qqplot(-np.log10(df_plot), 'unif', 'real', xlabel='Expected $-\log_{10}(p)$', ylabel='Observed $-\log_{10}(p)$', s=5, linewidth=0)
ax.set_xlim([0, 5])